src/multiplayer.cpp

```diff
@@ -4,6 +4,142 @@
 #include "engine.h"
 #include "multiplayer_internal.h"
 
+namespace replication
+{
+    Item::Item(ControlBlock& control, const Settings& settings)
+    : controller{ control }
+    {
+        controller.add(*this, settings);
+    }
+
+    Item::~Item() = default;
+
+    void Item::setId(Key, uint16_t id)
+    {
+        replication_id = id;
+    }
+
+    ControlBlock& Item::getController(Key) const
+    {
+        return controller;
+    }
+
+    uint16_t Item::getId(Key) const
+    {
+        return replication_id;
+    }
+
+    void Item::setDirty()
+    {
+        controller.setDirty(*this);
+    }
+
+    void ControlBlock::add(Item& item, const Item::Settings& settings)
+    {
+        assert(items.size() < static_cast<size_t>(~controlled_flag));
+        item.setId({}, static_cast<uint16_t>(items.size()));
+        items.emplace_back(&item);
+        min_update_interval.emplace_back(settings.min_delay);
+        time_since_last_update.emplace_back(0.f);
+        if (dirty.size() * sizeof(size_t) < items.size())
+            dirty.resize((items.size() + sizeof(size_t) - 1) / sizeof(size_t));
+
+#ifdef DEBUG
+            names.emplace_back(settings.name);
+#endif
+    }
+
+    bool ControlBlock::isDirty(const Item& item) const
+    {
+        assert(&item.getController({}) == this);
+        return isDirty(item.getId({}));
+    }
+
+    void ControlBlock::setDirty(const Item& item)
+    {
+        assert(&item.getController({}) == this);
+        setDirty(item.getId({}));
+    }
+
+    size_t ControlBlock::send(sf::Packet& packet, bool everything, float delta /* = 0.f */)
+    {
+        size_t sent{};
+
+        // Update all deltas.
+        for (auto& value : time_since_last_update)
+            value += delta;
+
+        if (!everything)
+        {
+            for (auto dirty_batch = 0; dirty_batch < dirty.size(); ++dirty_batch)
```

Comment:
Not sure I like this. My method with the single direction linked list was simpler and more efficient, compared to these packed dirty bits.

Reply:
I'm not sure about yours being more efficient: linked lists involve a ton of pointer chasing which are rarely a prefetcher's favorite, and it required an extra pointer on the replicated members themselves, so the memory footprint was also higher as well.

However, again, measurement is king.

Reply:
> My point on being more efficient is, the head pointer will usually point to nullptr

Interesting.
In my tests, where I converted a few of the fields on the spaceship, there's almost always 'something' that changed among all the replicated values, so from what I've it looked like the list would never really be empty...
And I figured it'd get worse once alll the fields would be converted.

Reply:
Ah, see where our difference is then, you want to convert all the fields, I just want to convert the ones that almost never change.